DREAM/Optimization/tsgGradientDescent.cpp

```diff
@@ -0,0 +1,130 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * UT-BATTELLE, LLC AND THE UNITED STATES GOVERNMENT MAKE NO REPRESENTATIONS AND DISCLAIM ALL WARRANTIES, BOTH EXPRESSED AND
+ * IMPLIED. THERE ARE NO EXPRESS OR IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE, OR THAT THE USE OF
+ * THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_GRADIENT_DESCENT_CPP
+#define __TASMANIAN_GRADIENT_DESCENT_CPP
+
+#include "tsgGradientDescent.hpp"
+
+namespace TasOptimization {
+
+GradientDescentState::GradientDescentState(const std::vector<double> &x0, const double lambda0) :
+        num_dimensions((int) x0.size()), adaptive_stepsize(lambda0), x(x0) {};
+
+OptimizationStatus GradientDescent(const ObjectiveFunctionSingle &func, const GradientFunctionSingle &grad,
+                                   const ProjectionFunctionSingle &proj, const double increase_coeff,
+                                   const double decrease_coeff, const int num_iterations, const double tolerance,
+                                   GradientDescentState &state) {
+
+    OptimizationStatus status;
+    status.performed_iterations = 0;
+    status.residual = std::numeric_limits<double>::max();
+
+    int num_dimensions = state.num_dimensions;
+    std::vector<double> &x = state.getXRef();
+    std::vector<double> x0(x), gx0(grad(x0)), gx(num_dimensions), xStep(num_dimensions);
+    double fx0(func(x0)), fxStep(0);
+
+    while(status.performed_iterations < num_iterations) {
+        // Find the next stepsize/candidate point by making sure it satisfies the well-known descent inequality (see γ_u in the
+        // reference paper associated with this function).
+        double lhs = 0;
+        double rhs = 0;
+        do {
+            if (status.performed_iterations >= num_iterations)
+                break;
+            for (int j=0; j<num_dimensions; j++)
+                xStep[j] = x0[j] - gx0[j] * state.adaptive_stepsize;
+            xStep = proj(xStep);
+            fxStep = func(xStep);
+            lhs += fxStep - fx0;
+            for (int j=0; j<num_dimensions; j++) {
+                double delta = xStep[j] - x0[j];
+                rhs += delta * delta / (2.0 * state.adaptive_stepsize);
+                lhs -= gx0[j] * delta;
+            }
+            state.adaptive_stepsize /= decrease_coeff;
```

Comment:
If `lhs < rhs` do you still reduce the `adaptive_stepsize`?

Reply:
Yes. This is okay because the stepsize is then multiplied by `decrease_coeff` right after the termination of the while-loop.

To make the logic of this approach more clear, it's helpful to consider all the cases that occur and what we expect the adaptive stepsize to be at the end of one iteration of the main while loop. 

Specifically, denoting `stepsize0` to be stepsize at the beginning of an iteration of the `while(status.performed_iterations < num_iterations)` loop, we expect:

**Case 1 - A break occurs inside the do-while loop using the `(lhs > rhs)` condition**
`adaptive_stepsize = stepsize0 / (decrease_coeff ^ performed_iterations)`

**Case 2 - A break occurs due to the `status.residual <= tolerance` condition**
`adaptive_stepsize = stepsize0 / (decrease_coeff ^ [performed_iterations - 1])`

**Case 3 - The optimistic update is performed**
`adaptive_stepsize = stepsize0 / (decrease_coeff ^ [performed_iterations - 1] * increase_coeff)`

We can discuss more if this is confusing.